### ISSUE TITLE
GRPC Web Support

### ISSUE THREAD

**Issue by jakepoz** [containous/traefik#4210]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

We would like to support some services that use the new GRPC-Web protocol.

https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md

Currently it seems that only Envoy is capable of proxying GRPC-Web traffic.

**Comment by jvmlet** [comment]
Is there any progress with this feature ? Supporting grpc web would be a great addition for traefik portfolio.

**Comment by Jarzamendia** [comment]
Yeah, we are now using an Envoy proxy to support gRPC Web with Traefki like a Edge proxy. It's a good update to remove Envoy proxy and only use traefik!

**Comment by fedojo** [comment]
Hey! 

Is it related with Traefik in version 1 or 2? Is there any option to use gRPC web in Traefik 2? 
Currently we are using Envoy but it would be great to move to Traefik!

Thanks in advance for any tip in this area.

**Comment by TannerYoung** [comment]
@Jarzamendia 

Do you have any documentation on how you're doing this exactly? I'm not familiar at all with envoy, and would greatly appreciate any documentation, or example (especially docker-compose)

Thanks

**Comment by nadilas** [comment]
I have been meaning to change over to using Traefik for a while now and today I managed to re-implement all our routers. Without checking grpc-web support in advance... So I began experimenting in our test environment and managed to get past the inital CORS error. Which means now the backend cannot handle the request (500 - Internal Server Error) obviously.

I can imagine two ways how I can proceed with this, but not knowing the current status of this feature request I am quite unsure:
1. One option would be to add grpc-web support within the backend service and create a separate route for grpc-web requests. Seems reasonable although I'm not certain if server streaming response would work.
2. An even nicer solution would be to have a grpc-web transformer middleware, which add to a separate router and point it directly at the original backend code.

@ldez Before I do anything or abandon Traefik for now, I wanted to ask you what the status on this topic is? Is there any progress being made? Can you contribute somehow?

**Comment by Jarzamendia** [comment]
I did a lab on this a while ago, I'm going to send it to git to show you. I haven't done any troughput tests in production yet. Using only Traefik, I have had good results in the infrastructure I use. I still have to test Envoy better.
My lab is in: https://github.com/Jarzamendia/traefik/tree/master/1.7.x/GRPC/GRPC-WEB

**Comment by nadilas** [comment]
Just thinking about this, can we not make use of this [grpcwebproxy](https://github.com/improbable-eng/grpc-web/blob/cd3cae9a5c1c1f4ea7b15ce13e976cf74d8b954c/go/grpcwebproxy/main.go#L65) within https://github.com/containous/traefik/blob/829649e905700ba6f80313e466362c0dd65f32a4/pkg/server/service/service.go#L170 ?

Pretty sure there's more to this, but it does seem like a good start? Granted that would require an additional parameter on either `Service` or `ServersLoadBalancer` to turn the feature on/off which might not be ideal.

**Comment by saeidakbari** [comment]
Is there any update for this feature?

**Comment by JaneJeon** [comment]
Almost 3 years since the issue was opened. Looks like this one isn't getting looked at...

**Comment by The-Judge** [comment]
+1

**Comment by thientu** [comment]
This issue is still open

**Comment by kahirokunn** [comment]
+1
